### PR TITLE
Apply fast-forward update to default branch of forked repo prior to submission

### LINKED ISSUE #233: wingetcreate can't create a reference when the forked repo has >= 254 branches
### Brief description of your issue

wingetcreate (latest) can no longer create PR's successfully. The token which is used is still valid.
Looking at the error, my assumption is that [this](https://github.com/microsoft/winget-create/blob/10dc9bde27fa4dc2f7c231a2aace6ccbc6e876d3/src/WingetCreateCore/Common/GitHub.cs#L314) call fails.

```logs
Telemetry Settings
------------------
The Windows Package Manager Manifest Creator collects usage data in order to improve your experience.
The data is anonymous and collected only by Microsoft.
By default, telemetry is enabled but can be disabled by running `wingetcreate settings` and editing your settings file.

Manifest validation succeeded: True

Submitting pull request for manifest...

Octokit.NotFoundException: Not Found
   at Octokit.Connection.HandleErrors(IResponse response)
   at Octokit.Connection.RunRequest(IRequest request, CancellationToken cancellationToken)
   at Octokit.Connection.Run[T](IRequest request, CancellationToken cancellationToken)
   at Octokit.ApiConnection.Post[T](Uri uri, Object data, String accepts, String contentType, CancellationToken cancellationToken)
   at Microsoft.WingetCreateCore.Common.GitHub.<>c__DisplayClass[19](https://github.com/JanDeDobbeleer/oh-my-posh/runs/5403344265?check_suite_focus=true#step:3:19)_0.<<SubmitPRAsync>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Polly.AsyncPolicy.<>c__DisplayClass40_0.<<ImplementationAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Polly.Retry.AsyncRetryEngine.ImplementationAsync[TResult](Func`3 action, Context context, CancellationToken cancellationToken, ExceptionPredicates shouldRetryExceptionPredicates, ResultPredicates`1 shouldRetryResultPredicates, Func`5 onRetryAsync, Int[32](https://github.com/JanDeDobbeleer/oh-my-posh/runs/5403344265?check_suite_focus=true#step:3:32) permittedRetryCount, IEnumerable`1 sleepDurationsEnumerable, Func`4 sleepDurationProvider, Boolean continueOnCapturedContext)
   at Polly.AsyncPolicy.ExecuteAsync(Func`3 action, Context context, CancellationToken cancellationToken, Boolean continueOnCapturedContext)
   at Microsoft.WingetCreateCore.Common.GitHub.SubmitPRAsync(String packageId, String version, Dictionary`2 contents, Boolean submitToFork)
   at Microsoft.WingetCreateCore.Common.GitHub.SubmitPRAsync(String packageId, String version, Dictionary`2 contents, Boolean submitToFork)
   at Microsoft.WingetCreateCLI.Commands.BaseCommand.GitHubSubmitManifests(Manifests manifests)
   at Microsoft.WingetCreateCLI.Commands.SubmitCommand.SubmitManifest()
   at Microsoft.WingetCreateCLI.Commands.SubmitCommand.Execute()
   at Microsoft.WingetCreateCLI.Program.Main(String[] args)
{"message":"Not Found","documentation_url":"https://docs.github.com/rest/reference/git#create-a-reference"}

Error: Process completed with exit code 1.
```

### Steps to reproduce

```powershell
Invoke-WebRequest 'https://aka.ms/wingetcreate/latest/self-contained' -OutFile wingetcreate.exe
# Create the PR
./wingetcreate.exe submit --token $Token $Version
```

### Expected behavior

It creates a PR sucessfully

### Actual behavior

It errors

### Environment

```shell
GitHub Actions windows-latest
```

### LINKED ISSUE #234: Winget Create Error
The following error message appeared when submitting an update to an existing package (gdiObjects.Perspective)
-----------
Manifest saved to C:\WINDOWS\system32\manifests\g\gdiObjects\Perspective\3.1.2203.2

Manifest validation succeeded: True

Submitting pull request for manifest...

Octokit.NotFoundException: Not Found
   at Octokit.Connection.HandleErrors(IResponse response) in /home/runner/work/octokit.net/octokit.net/Octokit/Http/Connection.cs:line 700
   at Octokit.Connection.RunRequest(IRequest request, CancellationToken cancellationToken) in /home/runner/work/octokit.net/octokit.net/Octokit/Http/Connection.cs:line 683
   at Octokit.Connection.Run[T](IRequest request, CancellationToken cancellationToken) in /home/runner/work/octokit.net/octokit.net/Octokit/Http/Connection.cs:line 668
   at Octokit.ApiConnection.Post[T](Uri uri, Object data, String accepts, String contentType, CancellationToken cancellationToken) in /home/runner/work/octokit.net/octokit.net/Octokit/Http/ApiConnection.cs:line 293
   at Microsoft.WingetCreateCore.Common.GitHub.<>c__DisplayClass19_0.<<SubmitPRAsync>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Polly.AsyncPolicy.<>c__DisplayClass40_0.<<ImplementationAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Polly.Retry.AsyncRetryEngine.ImplementationAsync[TResult](Func`3 action, Context context, CancellationToken cancellationToken, ExceptionPredicates shouldRetryExceptionPredicates, ResultPredicates`1 shouldRetryResultPredicates, Func`5 onRetryAsync, Int32 permittedRetryCount, IEnumerable`1 sleepDurationsEnumerable, Func`4 sleepDurationProvider, Boolean continueOnCapturedContext)
   at Polly.AsyncPolicy.ExecuteAsync(Func`3 action, Context context, CancellationToken cancellationToken, Boolean continueOnCapturedContext)
   at Microsoft.WingetCreateCore.Common.GitHub.SubmitPRAsync(String packageId, String version, Dictionary`2 contents, Boolean submitToFork)
   at Microsoft.WingetCreateCore.Common.GitHub.SubmitPRAsync(String packageId, String version, Dictionary`2 contents, Boolean submitToFork)
   at Microsoft.WingetCreateCLI.Commands.BaseCommand.GitHubSubmitManifests(Manifests manifests)
   at Microsoft.WingetCreateCLI.Commands.UpdateCommand.ExecuteManifestUpdate(List`1 latestManifestContent, CommandExecutedEvent commandEvent)
   at Microsoft.WingetCreateCLI.Commands.UpdateCommand.Execute()
   at Microsoft.WingetCreateCLI.Program.Main(String[] args)
{"message":"Not Found","documentation_url":"https://docs.github.com/rest/reference/git#create-a-reference"}

### PR DESCRIPTION
**Pull request by ryfu-msft**
Resolves #233 
Resolves #234 

GitHub API will throw a NotFoundException when trying to create a reference if the default branch of the forked 'winget-pkgs' repository is not up to date with the latest upstream commits. 

The changes in this PR resolve these issues by checking for any upstream commits and updating the fork's default branch prior to submission. The update will only work successfully if the update is only a fast-forward update.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-create/pull/235)

### CODE REVIEW

File: src/WingetCreateCore/Common/GitHub.cs [hunk 1]
```diff
@@ -366,6 +369,24 @@ private async Task<PullRequest> SubmitPRAsync(string packageId, string version,
...
+        /// Checks if the provided forked repository is behind on upstream commits and updates the default branch with the fetched commits. Update can only be a fast-forward update.
+        /// </summary>
+        /// <param name="forkedRepo"><see cref="Repository"/>Forked repository to be updated.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        private async Task UpdateForkedRepoWithUpstreamCommits(Repository forkedRepo)
+        {
+            var upstream = forkedRepo.Parent;
+            var compareResult = await this.github.Repository.Commit.Compare(upstream.Id, upstream.DefaultBranch, $"{forkedRepo.Owner.Login}:{forkedRepo.DefaultBranch}");
+
+            // Check to ensure that the update is only a fast-forward update.
```
Comment (yao-msft) [inline src/WingetCreateCore/Common/GitHub.cs:382]: >// Check to ensure that the update is only a fast-forward update. [](http://example.com/codeflow?start=12&length=65)

suggest: for cases that cannot be ff'ed, throw a custom exception so the caller can catch it and display a corresponding error message

Comment (ryfu-msft) [reply src/WingetCreateCore/Common/GitHub.cs:391]: Added custom exception to display the correct error message to the user.

File: src/WingetCreateCore/Common/GitHub.cs [hunk 2]
```diff
@@ -306,12 +307,23 @@ private async Task<PullRequest> SubmitPRAsync(string packageId, string version,
...
+
             try
             {
                 var retryPolicy = Policy.Handle<ApiException>().WaitAndRetryAsync(3, i => TimeSpan.FromSeconds(i));
                 await retryPolicy.ExecuteAsync(async () =>
                 {
-                    await this.github.Git.Reference.Create(repo.Id, new NewReference($"refs/{newBranchNameHeads}", upstreamMasterSha));
+                    try
+                    {
+                        Reference reference = await this.github.Git.Reference.Create(repo.Id, new NewReference($"refs/{newBranchNameHeads}", upstreamMasterSha));
```
Comment (yao-msft) [inline src/WingetCreateCore/Common/GitHub.cs:318]: >Reference reference [](http://example.com/codeflow?start=24&length=19)

seems like this local variable is assigned but not used?

**Review by yao-msft — APPROVED** [review]
:shipit:

File: src/WingetCreateCore/Common/GitHub.cs (hunk 2 above)
Comment (ryfu-msft) [reply src/WingetCreateCore/Common/GitHub.cs]: Removed unused local variable

**Pull request merged**